cwl/src/main/scala/cwl/CwlJsEncoder.scala

```diff
@@ -0,0 +1,91 @@
+package cwl
+
+import cats.instances.list._
+import cats.syntax.traverse._
+import cats.syntax.validated._
+import common.validation.ErrorOr._
+import common.validation.Validation._
+import wom.util.JsEncoder
+import wom.values.{WomFile, WomGlobFile, WomMaybeListedDirectory, WomMaybePopulatedFile, WomSingleFile, WomUnlistedDirectory, WomValue}
+
+import scala.collection.JavaConverters._
+
+class CwlJsEncoder extends JsEncoder {
+  /**
+    * Overrides encoding to also support wom file or directory values.
+    */
+  override def encode(value: WomValue): ErrorOr[AnyRef] = {
+    value match {
+      case file: WomFile => encodeFileOrDirectory(file)
+      case _ => super.encode(value)
+    }
+  }
+
+  /**
+    * Encodes a sequence of wom file or directory values.
+    */
+  def encodeFileOrDirectories(values: Seq[WomFile]): ErrorOr[Array[java.util.Map[String, AnyRef]]] = {
+    values.toList.traverse(encodeFileOrDirectory).map(_.toArray)
+  }
+
+  /**
+    * Encodes a wom file or directory value.
+    */
+  def encodeFileOrDirectory(value: WomFile): ErrorOr[java.util.Map[String, AnyRef]] = {
+    value match {
+      case directory: WomUnlistedDirectory => encodeDirectory(WomMaybeListedDirectory(directory.value))
+      case file: WomSingleFile => encodeFile(WomMaybePopulatedFile(file.value))
+      case glob: WomGlobFile => s"Glob file javascript encoding is not supported: $glob".invalidNel
+      case directory: WomMaybeListedDirectory => encodeDirectory(directory)
+      case file: WomMaybePopulatedFile => encodeFile(file)
+    }
+  }
+
+  /**
+    * Encodes a wom file.
+    */
+  def encodeFile(file: WomMaybePopulatedFile): ErrorOr[java.util.Map[String, AnyRef]] = {
+    val lifted: ErrorOr[Map[String, Option[AnyRef]]] = Map(
+      "class" -> validate(Option("File")),
+      "location" -> validate(Option(file.value)),
+      "path" -> validate(Option(file.value)),
+      "basename" -> validate(Option(File.basename(file.value))),
+      "dirname" -> validate(Option(File.dirname(file.value))),
+      "nameroot" -> validate(Option(File.nameroot(file.value))),
+      "nameext" -> validate(Option(File.nameext(file.value))),
+      "checksum" -> validate(file.checksumOption),
+      "size" -> validate(file.sizeOption.map(Long.box)),
+      "secondaryFiles" -> encodeFileOrDirectories(file.secondaryFiles).map(Option(_)),
+      "format" -> validate(file.formatOption),
+      "contents" -> validate(file.contentsOption)
+    ).sequence
+
+    flattenToJava(lifted)
+  }
+
+  /**
+    * Encodes a wom directory.
+    */
+  def encodeDirectory(directory: WomMaybeListedDirectory): ErrorOr[java.util.Map[String, AnyRef]] = {
+    val lifted: ErrorOr[Map[String, Option[AnyRef]]] = Map(
+      "class" -> validate(Option("Directory")),
+      "location" -> validate(directory.valueOption),
+      "path" -> validate(Option(directory.value)),
```

Comment:
Yes. I've punted this for now as it will take a bit suddenly tell JES/SFS "you only get to pick where files go when there isn't this user provided 'path'".

Currently, backends look at the `WomFile.value` as a full path to the original file, make a copy of the `WomFile` substituting in a localized path for the backend, and then call `instantiateCommand` using the `WomFile.value` with the modified path.

Using user supplied paths will be a bit of rewiring, and a new ticket.